python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,601 @@
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    Singleton,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+EPOCH = datetime.datetime.utcfromtimestamp(0)
+
+"""
+Iceberg literal is wrapper class used in expressions, which return unbound predicates
+It's being organized as below
+Literal
+|-- AboveMax
+|-- BelowMin
+|-- BaseLiteral
+    |-- FixedLiteral
+    |-- BinaryLiteral
+    |-- ComparableLiteral
+        |-- BooleanLiteral
+        |-- IntegerLiteral
+        |-- LongLiteral
+        |-- FloatLiteral
+        |-- DoubleLiteral
+        |-- DateLiteral
+        |-- StringLiteral
+        |-- TimeLiteral
+        |-- TimestampLiteral
+        |-- DecimalLiteral
+        |-- UUIDLiteral
+"""
+
+
+class Literal:
+    def to(self, type_var: PrimitiveType):
+        raise NotImplementedError()
+
+    def __str__(self):
+        return type(self).__name__
+
+
+class BaseLiteral(Literal):
+    """Base literal which has a value and can be converted between types"""
+
+    def __init__(self, repr_string: str, value):
+        self._repr_string = repr_string
+        if value is None:
+            raise TypeError("Cannot set value of BaseLiteral to None")
+        self._value = value
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    @property
+    def value(self):
+        return self._value
+
+    def __repr__(self):
+        return self._repr_string
+
+    def __str__(self):
+        return str(self._value)
+
+
+@singledispatch
+def literal(value: Any) -> BaseLiteral:
+    """
+    A generic Literal factory to construct an iceberg Literal based on python primitive data type
+    using dynamic overloading
+
+    Args:
+        value(python primitive type): the value to be associated with literal
+
+    Example:
+        from iceberg.expressions.literals import literal
+        >>> literal(123)
+        <IntegerLiteral(123)>v
+    """
+    raise TypeError(f"Unimplemented Type Literal for value: {value}")
+
+
+@literal.register
+def _(value: bool):
+    return BooleanLiteral(value)
+
+
+@literal.register
+def _(value: int):
+    """
+    Upgrade to long if python int is outside the JAVA_MIN_INT and JAVA_MAX_INT
+    """
+    if value < IntegerType.min or value > IntegerType.max:
+        return LongLiteral(value)
+    return IntegerLiteral(value)
+
+
+@literal.register
+def _(value: float):
+    """
+    Upgrade to double if python float is outside the JAVA_MIN_FLOAT and JAVA_MAX_FLOAT
+    """
+    if value < FloatType.min or value > FloatType.max:
+        return DoubleLiteral(value)
+    return FloatLiteral(value)
+
+
+@literal.register
+def _(value: str):
+    return StringLiteral(value)
+
+
+@literal.register
+def _(value: uuid.UUID):
+    return UUIDLiteral(value)
+
+
+@literal.register
+def _(value: bytes):
+    return FixedLiteral(value)
```

Comment:
I think that both `bytes` and `bytearray` should produce a `BinaryLiteral` that contains `bytes`. Iceberg should not hold on to `bytearray` values because they are mutable, so this should be immediately converted in the factory method.